grcs/0020-knowledge-graph.md

```diff
@@ -373,100 +369,103 @@ An op represents a single atomic operation that produces or modifies knowledge.
 ```proto
 message Op {
   OpType type = 1;
-  Triple triple = 2;
-  Entity entity = 3;
-  Relation relation = 4;
-  repeated Triple triples = 5;
+  Entity entity = 2;
+  Relation relation = 3;
+  Property property = 4;
 }
-```
+``` 
 
 ```proto
 message Relation {
-  string id = 1;
-  string type = 2;
-  string from_entity = 3;
-  string to_entity = 4;
-  string index = 5;
+  bytes id = 1;
+  bytes type = 2;
+  bytes from_entity = 3;
+  bytes from_property = 4;
+  bytes to_entity = 5;
+  bytes to_space = 6;
+  bytes entity = 7;
+  string index = 8;
 }
 ```
 
 ```proto
-message Entity {
-  string id = 1;
-  repeated string types 2;
+message Property {
+  bytes id = 1;
+  bytes type = 2;
+  string name = 3;
 }
-```
-
-| Name                   | Value |
-| ---------------------- |-------|
-| Set triple             | 1     |
-| Delete triple          | 2     |
-| Set triple batch       | 3     |
-| Delete entity          | 4     |
-| Create relation        | 5     |
-| Delete relation        | 6     |
-
+``` 
 
 
 ```proto
 enum OpType {
-  SET_TRIPLE = 1;
-  DELETE_TRIPLE = 2;
-  SET_TRIPLE_BATCH = 3;
-  DELETE_ENTITY = 4;
-  CREATE_RELATION = 5;
-  DELETE_RELATION = 6;
+  CREATE_ENTITY = 1;
+  UPDATE_ENTITY = 2;
+  DELETE_ENTITY = 3;
+  CREATE_RELATION = 4;
+  DELETE_RELATION = 5;
+  REORDER_RELATION = 6;
+  UNSET_PROPERTIES = 7;
+  CREATE_PROPERTY = 8;
+  ARCHIVE_PROPERTY = 9;
+  MOVE_ENTITY = 10;
+  MERGE_ENTITIES = 11;
+  BRANCH_ENTITY = 12;
```

Comment:
Byron:

Should we add a spec + op for importing a file?

What do we think about adding a SET_SPACE_ENTITY op? This would allow us to unify the ids for the space and its front page instead of having two different ids. This has been a point of confusion in the past and not sure the benefits outweighs the awkwardness.

let op = {
   type: OpType.SET_SPACE_ENTITY,
   entity: {
       id: "id-of-entity-to-be-space-front-page"
   }
}
Any subsequent SET_SPACE_ENTITY ops would upsert to ensure there's only ever one at a time at the data service level.